File: ddns/provider/dnspod.py
# coding=utf-8
"""
DNSPOD API
@doc: https://docs.dnspod.cn/api/
@author: NewFuture
"""

from ._base import BaseProvider, TYPE_FORM


class DnspodProvider(BaseProvider):
    """
    DNSPOD API
    DNSPOD 接口解析操作库
    """

    endpoint = "https://dnsapi.cn"
    content_type = TYPE_FORM

    DefaultLine = "默认"

    def _request(self, action, extra=None, **params):
        # type: (str, dict | None, **(str | int | bytes | bool | None)) -> dict
        """
        发送请求数据

        Send request to DNSPod API.
        Args:
            action (str): API 动作/Action
            extra (dict|None): 额外参数/Extra params
            params (dict): 其它参数/Other params
        Returns:
            dict: 响应数据/Response data
        """
        # 过滤掉None参数
        if extra:
            params.update(extra)
        params = {k: v for k, v in params.items() if v is not None}
        params.update({"login_token": "{0},{1}".format(self.id, self.token), "format": "json"})
        data = self._http("POST", "/" + action, body=params)
        if data and data.get("status", {}).get("code") == "1":  # 请求成功
            return data
        else:  # 请求失败
            error_msg = "Unknown error"
            if data and isinstance(data, dict):
                error_msg = data.get("status", {}).get("message", "Unknown error")
            self.logger.warning("DNSPod API error: %s", error_msg)
            return data

    def _create_record(self, zone_id, subdomain, main_domain, value, record_type, ttl, line, extra):
        # type: (str, str, str, str, str, int | str | None, str | None, dict) -> bool
        """https://docs.dnspod.cn/api/add-record/"""
        res = self._request(
            "Record.Create",
            extra=extra,
            domain_id=zone_id,
            sub_domain=subdomain,
            value=value,
            record_type=record_type,
            record_line=line or self.DefaultLine,
            ttl=ttl,
        )
        record = res and res.get("record")
        if record:  # 记录创建成功
            self.logger.info("Record created: %s", record)
            return True
        else:  # 记录创建失败
            self.logger.error("Failed to create record: %s", res)
        return False

    def _update_record(self, zone_id, old_record, value, record_type, ttl, line, extra):
        # type: (str, dict, str, str, int | str | None, str | None, dict) -> bool
        """https://docs.dnspod.cn/api/modify-records/"""
        record_line = (line or old_record.get("line") or self.DefaultLine).replace("Default", "default")
        res = self._request(
            "Record.Modify",
            domain_id=zone_id,
            record_id=old_record.get("id"),
            sub_domain=old_record.get("name"),
            record_type=record_type,
            value=value,
            record_line=record_line,
            extra=extra,
        )
        record = res and res.get("record")
        if record:  # 记录更新成功
            self.logger.debug("Record updated: %s", record)
            return True
        else:  # 记录更新失败
            self.logger.error("Failed to update record: %s", res)
            return False

    def _query_zone_id(self, domain):
        # type: (str) -> str | None
        """查询域名信息 https://docs.dnspod.cn/api/domain-info/"""
        res = self._request("Domain.Info", domain=domain)
        if res and isinstance(res, dict):
            return res.get("domain", {}).get("id")
        return None

    def _query_record(self, zone_id, subdomain, main_domain, record_type, line, extra):
        # type: (str, str, str, str, str | None, dict) -> dict | None
        """查询记录 list 然后逐个查找 https://docs.dnspod.cn/api/record-list/"""
        res = self._request("Record.List", domain_id=zone_id, sub_domain=subdomain, record_type=record_type, line=line)
        # length="3000"
        records = res.get("records", [])
        n = len(records)
        if not n:
            self.logger.warning("No record found for [%s] %s<%s>(line: %s)", zone_id, subdomain, record_type, line)
            return None
        if n > 1:
            self.logger.warning("%d records found for %s<%s>(%s):\n %s", n, subdomain, record_type, line, records)
            return next((r for r in records if r.get("name") == subdomain), None)
        return records[0]

File: rust/src/provider/dnspod.rs
use std::collections::BTreeMap;

use serde_json::{Map, Value, json};

use crate::error::{Error, Result};
use crate::http::{Method, form_encode};

use super::base::{CrudProvider, ProviderContext, RecordRequest, json_parameters, value_to_string};

pub struct DnspodProvider<'a> {
    context: ProviderContext<'a>,
    zones: BTreeMap<String, String>,
    default_line: &'static str,
}

impl<'a> DnspodProvider<'a> {
    pub fn new(context: ProviderContext<'a>) -> Result<Self> {
        Self::with_settings(context, "默认")
    }

    pub fn global(context: ProviderContext<'a>) -> Result<Self> {
        Self::with_settings(context, "default")
    }

    fn with_settings(context: ProviderContext<'a>, default_line: &'static str) -> Result<Self> {
        if context.id.is_empty() {
            return Err(Error::Config("DNSPod id must be configured".to_owned()));
        }
        if context.token.is_empty() {
            return Err(Error::Config("DNSPod token must be configured".to_owned()));
        }
        Ok(Self {
            context,
            zones: BTreeMap::new(),
            default_line,
        })
    }

    fn api(&self, action: &str, parameters: Map<String, Value>) -> Result<Value> {
        let mut parameters = parameters
            .into_iter()
            .filter_map(|(key, value)| value_to_string(&value).map(|value| (key, value)))
            .collect::<BTreeMap<_, _>>();
        parameters.insert(
            "login_token".to_owned(),
            format!("{},{}", self.context.id, self.context.token),
        );
        parameters.insert("format".to_owned(), "json".to_owned());
        let headers = BTreeMap::from([(
            "content-type".to_owned(),
            "application/x-www-form-urlencoded".to_owned(),
        )]);
        let response = self.context.send_json(
            Method::Post,
            &format!("/{action}"),
            &BTreeMap::new(),
            Some(form_encode(&parameters)),
            headers,
        )?;
        let status_code = response
            .pointer("/status/code")
            .and_then(value_to_string)
            .unwrap_or_else(|| "unknown".to_owned());
        let zone_candidate_miss =
            action == "Domain.Info" && matches!(status_code.as_str(), "6" | "7" | "8");
        if status_code == "1"
            || (action == "Record.List" && status_code == "10")
            || zone_candidate_miss
        {
            return Ok(response);
        }
        let message = response
            .pointer("/status/message")
            .and_then(Value::as_str)
            .unwrap_or("unknown error");
        Err(Error::Provider(format!(
            "DNSPod API error {status_code}: {}",
            self.context.logger.mask(message)
        )))
    }
}

impl CrudProvider for DnspodProvider<'_> {
    fn context(&self) -> &ProviderContext<'_> {
        &self.context
    }

    fn zone_cache(&mut self) -> &mut BTreeMap<String, String> {
        &mut self.zones
    }

    fn query_zone_id(&mut self, domain: &str) -> Result<Option<String>> {
        let response = self.api(
            "Domain.Info",
            Map::from_iter([("domain".to_owned(), json!(domain))]),
        )?;
        Ok(response.pointer("/domain/id").and_then(value_to_string))
    }

    fn query_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        _main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<Option<Value>> {
        let mut parameters = Map::from_iter([
            ("domain_id".to_owned(), json!(zone_id)),
            ("sub_domain".to_owned(), json!(subdomain)),
            ("record_type".to_owned(), json!(request.record_type)),
        ]);
        if let Some(line) = request.line {
            parameters.insert("line".to_owned(), json!(line));
        }
        let response = self.api("Record.List", parameters)?;
        let records = response
            .get("records")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        Ok(if records.len() <= 1 {
            records.into_iter().next()
        } else {
            records
                .into_iter()
                .find(|record| record.get("name").and_then(Value::as_str) == Some(subdomain))
        })
    }

    fn create_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        _main_domain: &str,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let mut parameters = Map::from_iter([
            ("domain_id".to_owned(), json!(zone_id)),
            ("sub_domain".to_owned(), json!(subdomain)),
            ("value".to_owned(), json!(request.address)),
            ("record_type".to_owned(), json!(request.record_type)),
            (
                "record_line".to_owned(),
                json!(request.line.unwrap_or(self.default_line)),
            ),
            ("ttl".to_owned(), json!(request.ttl)),
        ]);
        // DNSPod applies extras last, before omitting null/non-scalar form values.
        parameters.extend(json_parameters(request));
        let response = self.api("Record.Create", parameters)?;
        if response.get("record").is_some() {
            Ok(())
        } else {
            Err(Error::Provider(format!(
                "DNSPod failed to create record: {response}"
            )))
        }
    }

    fn update_record(
        &mut self,
        zone_id: &str,
        record: &Value,
        request: &RecordRequest<'_>,
    ) -> Result<()> {
        let record_line = request
            .line
            .or_else(|| record.get("line").and_then(Value::as_str))
            .unwrap_or(self.default_line)
            .replace("Default", "default");
        let mut parameters = Map::from_iter([
            ("domain_id".to_owned(), json!(zone_id)),
            ("record_id".to_owned(), json!(record.get("id"))),
            ("sub_domain".to_owned(), json!(record.get("name"))),
            ("record_type".to_owned(), json!(request.record_type)),
            ("value".to_owned(), json!(request.address)),
            ("record_line".to_owned(), json!(record_line)),
            ("ttl".to_owned(), json!(request.ttl)),
        ]);
        parameters.extend(json_parameters(request));
        let response = self.api("Record.Modify", parameters)?;
        if response.get("record").is_some() {
            Ok(())
        } else {
            Err(Error::Provider(format!(
                "DNSPod failed to update record: {response}"
            )))
        }
    }
}
